spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/integration/outbound/PubSubMessageHandler.java

```diff
@@ -16,47 +16,28 @@
 
 package org.springframework.cloud.gcp.pubsub.integration.outbound;
 
-import com.google.cloud.pubsub.spi.v1.Publisher;
-import com.google.pubsub.v1.PubsubMessage;
-import com.google.pubsub.v1.TopicName;
 import java.io.IOException;
-import org.springframework.cloud.gcp.pubsub.integration.converters.SimpleMessageConverter;
+import org.springframework.cloud.gcp.pubsub.core.PubSubTemplate;
 import org.springframework.integration.handler.AbstractMessageHandler;
 import org.springframework.messaging.Message;
-import org.springframework.messaging.converter.MessageConversionException;
-import org.springframework.messaging.converter.MessageConverter;
 
 /**
- * Converts from internal Spring message to GCP Pub/Sub message and publishes the Pub/Sub
- * message to a Pub/Sub topic.
+ * Sends messages to Google Cloud Pub/Sub by delegating to {@link PubSubTemplate}.
  *
  * @author João André Martins
  */
 public class PubSubMessageHandler extends AbstractMessageHandler {
 
-  private Publisher publisher;
-  private MessageConverter messageConverter;
+  private PubSubTemplate pubsubTemplate;
+  private String topic;
 
-  public PubSubMessageHandler(String projectId, String topicName) throws IOException {
-    this(projectId, topicName, new SimpleMessageConverter());
-  }
-
-  public PubSubMessageHandler(String projectId, String topicName,
-      MessageConverter messageConverter) throws IOException {
-    publisher = Publisher.defaultBuilder(TopicName.create(
-        projectId, topicName)).build();
-    this.messageConverter = messageConverter;
+  public PubSubMessageHandler(String projectId, String topic) throws IOException {
+    pubsubTemplate = new PubSubTemplate(null, projectId);
+    this.topic = topic;
```

Comment:
I guess my earlier point was also that, while it is possible to have a composite flow in SI like you guys described, with hooked `header-enrichers`, it's also possible to have a simpler flow where we can never set the headers (or it would be to cumbersome for users to do so).

For example, the following is a valid use case. It sounds like the SI adapter could support both cases simultaneously? Then we'd need to come with some sort of order for topic resolution, e.g., the topic passed in the header would have precedence over the one in the ctor/setter, or vice-versa.

```
  @Bean
  @ServiceActivator(inputChannel = "pubsubOutputChannel")
  public MessageHandler messageSender(@Qualifier("gcpProjectId") String projectId)
      throws IOException {
    return new PubSubMessageHandler(projectId, "test");
  }

  @Bean
  public MessageChannel pubsubOutputChannel() {
    return new PublishSubscribeChannel();
  }

  @MessagingGateway(defaultRequestChannel = "pubsubOutputChannel")
  public interface PubsubOutboundGateway {
    void sendToPubsub(String text);
  }
```